db/osqlcomm.c

```diff
@@ -6674,6 +6674,19 @@ int osql_process_packet(struct ireq *iq, unsigned long long rqid, uuid_t uuid,
 
         if (rc != 0) {
             if (err->errcode == OP_FAILED_UNIQ) {
+                if (iq->vfy_idx_track == 1 && iq->dup_key_insert == 1) {
+                    rc = ERR_UNCOMMITABLE_TXN;
+                    reqerrstr(iq, COMDB2_CSTRT_RC_DUP, "add key constraint "
+                                                   "duplicate key '%s' on "
+                                                   "table '%s' index %d",
+                          get_keynm_from_db_idx(iq->usedb, err->ixnum),
+                          iq->usedb->tablename, err->ixnum);
+                    err->blockop_num = step;
+                    err->ixnum = 0;
+                    err->errcode = ERR_UNCOMMITABLE_TXN;
```

Comment:
These error codes (`rc` and err->errcode`) do not need overwriting. I believe this would give us the correct rcode for the unique constraint violation.

Reply:
I decided to use `ERR_UNCOMMITABLE_TXN` because it makes it clear that the transaction failed on a verify because it was uncommittable. Using this return code causes the following line to be printed `[ERROR] Forced VERIFY-FAIL for uncommitable blocksql transaction`. This distinguishes it from a regular duplicate key constraint violation, where the transaction may still be committable on retry. The code that checks for duplicate updates on retries also uses this return code. 

What do you think?

Reply:
That makes sense. `ERR_UNCOMMITABLE_TXN` it is then!